src/interfaces/chain.cpp

```diff
@@ -344,6 +331,7 @@ class ChainImpl : public Chain
     }
     std::unique_ptr<Handler> handleNotifications(Notifications& notifications) override
     {
+        LOCK(::cs_main);
```

Comment:
3773b4159c37db993b9bb7c3f5d7dd7a51a39b09

Is this going to change in a follow up? Maybe add a comment explaining this lock?

Reply:
Adding a comment only for this method would seem obscure compare to other methods for future reviewers after refactoring, that's why it's in commit message "Lock cs_main in handleNotifications as it was relaying on its caller to do  so before", or do you have a different opinion ?

Reply:
re: https://github.com/bitcoin/bitcoin/pull/16426#discussion_r356338388

> Adding a comment only for this method would seem obscure compare to other methods for future reviewers after refactoring

I think this lock is a little different than the locks in the other Chain methods because it is not obvious why locking is needed for something that appears to just be a memory allocation. If this lock really is necessary, having a comment in the code would be helpful to me at least. Or if it is possible to drop the lock or move it closer to where it might be needed in the [`NotificationsHandlerImpl` constructor](https://github.com/bitcoin/bitcoin/blob/3773b4159c37db993b9bb7c3f5d7dd7a51a39b09/src/interfaces/chain.cpp#L44-L48)  that would be even better.

Reply:
Okay after thinking about this, by removing lock, we may miss block connections between end of rescan and registering with validation interface. That was the meaning of the comment I think, which is about consistency of chain viewed by the wallet. To avoid that I moved `handleNotifications` before rescan and after `cs_wallet` lock taking. That's way block connections notifications are going to be pending until rescan is done. We may have block processing duplicata but at least we should guarantee wallet consistency. 

Long-term solution is to unify both rescan and Chain::Notifications under same interface.